### ISSUE TITLE
map.getSource() returns undefined in onStyleLoad

### ISSUE THREAD

**Issue by khzaw** [alex3165/react-mapbox-gl#145]
I have a GEOJson `<Source id="point">` with hardcoded data (ideally will come from some other source) and accompanying `<Layer>` component inside my `<ReactMapboxGl>`. I am trying to do some animation upon loading of the map. But `map.getSource('point')` inside `onStyleLoad` is returning `undefined`.

My current workaround right now is to add a `setTimeout` and everything works fine by then.

```javascript
  onStyleLoad(map, e) {
    console.log(map.getSource('point'));   // this returns undefined
    setTimeout(_ => {
      console.log(map.getSource('point')); // this works
      // proceed to do other stuffs
    }, 1000);
  }
```

Here is my full code. Nothing really special. Initializing my variables and state in the constructor, rendering the map component. 

```javascript
import React, { Component } from 'react';
import ReactMapboxGl, { Layer, Source, Feature } from 'react-mapbox-gl';
import turf from '@turf/turf';

const MAPBOX_TOKEN = '<MAPBOX_TOKEN>';

class Map extends Component {
  constructor(props) {
    super(props);
    const path = turf.lineString([96.219456, 17.096823], [96.267761, 16.929063], [96.212375, 16.857859], [96.185887, 16.83685]);
    const point = turf.along(path, 0, 'miles');
    
    const pathSource = {
      type: 'geojson',
      data: path,
      maxzoom: 20
    };
    
    const pointSource = {
      type: 'geojson',
      data: point
    };
    
    this.state = {
      pathSource,
      pointSource,
      path,
      point
    };
    
    this.onStyleLoad = this.onStyleLoad.bind(this);
  }
  
  onStyleLoad(map, e) {
    console.log(map.getSource('point'));   // this returns undefined
    setTimeout(_ => {
      console.log(map.getSource('point')); // this works
      // proceed to do other stuffs
    }, 1000);
  }
  
  render() { 
    <ReactMapboxGl
      style="mapbox://styles/mapbox/dark-v9"
      accessToken={MAPBOX_TOKEN}
      center={[96.1587996, 16.7849606]}
      minZoom={10}
      zoom={[13.86]}
      bearing={1}
      containerStyle={{
        height: '100vh',
        width: '100vh'
      }}
      onStyleLoad={this.onStyleLoad}
    >
      <Source id={point} geoJsonSource={this.state.pointSource} />
      <Layer
        id="pointLayer"
        sourceId="point"
        type="circle"
        layout={{
        }}
        paint={{
          'circle-radius': 4,
          'circle-color': '#ffffff'
        }}
      >
      </Layer>     
    </ReactMapboxGl>
  }
}
```

I think maybe the map or the source layer is not fully loaded by the time `onStyleLoad` is fired. And is there a difference between `on('style.load')` and `on('load')`. I couldn't really find anything on the difference between the two in Mapbox GL docs. Should I be using `on('load')` here? Thanks!

**Comment by alex3165** [comment]
Hey it is normal because the `Source` component is not mounted when the  onStyleLoad function is executed. I have just added a `onSourceAdded` props which receive the source as a first argument to the `Source` component, you can use it to get the source when it has been added to the map.

**Comment by khzaw** [comment]
Ah … I see. `onSourceAdded` is just what I need. Thanks!

On Tue, 7 Mar 2017 at 17:40, Alexandre <notifications@github.com> wrote:

Hey it is normal because the Source component is not mounted when the
onStyleLoad function is executed. I have just added a onSourceAdded props
which receive the source as a first argument to the Source component, you
can use it to get the source when it has been added to the map.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/alex3165/react-mapbox-gl/issues/145#issuecomment-284670855>,
or mute the thread
